lib/vra/http.rb

```diff
@@ -0,0 +1,144 @@
+require 'net/http'
+
+module Vra
+  module Http
+
+
+    def self.execute(params)
+      request = Request.new(params)
+      response = request.call
+      while !response.final?
+        response = response.forward(request).call
+      end
+      fail error(response) unless response.success?
+      response
+    end
+
+
+    def self.error(response)
+      Error.from_response(response)
+    end
+
+    class Request
+      attr_reader :params
+
+      def initialize(params)
+        @params = params
+      end
+
+      def redirectable?
+        [:get, :head].include?(params[:method])
+      end
+
+      protected def new(new_params)
+        self.class.new(params.dup.merge(new_params))
+      end
+
+      def redirect_to(location)
+        new(url: location)
+      end
+
+      def see_other(location)
+        redirect_to(location).new(method: :get)
+      end
+
+      def call
+        uri = URI(params[:url]) or fail ":url required"
+
+        Net::HTTP.start(uri.host, uri.port,
+                        :use_ssl => uri.scheme == 'https') do |http|
+
+          request = http_request(params[:method], uri)
+          request.initialize_http_header(params[:headers] || {})
+          request.body = params[:payload] || ""
+
+          Response.new(http.request(request))
+        end
+      end
+
+
+      def http_request(method, uri)
+        type = {
+          get: Net::HTTP::Get,
+          head: Net::HTTP::Head,
+          post: Net::HTTP::Post
+        }.fetch(method, nil)
+
+        fail "Unknown HTTP method #{method}!" unless type
+
+        type.new(uri)
+      end
+
+    end
+
+
+    class Response
+      # For hiding the details of the HTTP response class
+      # so it can be swapped out easily
+      def initialize(response)
+        @response = response
+      end
+
+      def forward(request)
+        if redirect?
+          fail Http.error(self) unless request.redirectable?
+          request.redirect_to(location)
+        elsif see_other?
+          request.see_other(location)
+        else
+          request
+        end
+      end
+
+      def location()
```

Comment:
Unless there's a technical reason, let's be consistent with the method definitions, eliminating the `()` here as is done with other methods that have no arguments.